src/d3a/models/myco_matcher/pay_as_bid.py

```diff
@@ -0,0 +1,45 @@
+from d3a.constants import FLOATING_POINT_TOLERANCE
+
+
+class PayAsBidMatch:
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def sorting(obj, reverse_order=False):
+        if reverse_order:
+            # Sorted bids in descending order
+            return list(reversed(sorted(
+                obj.values(),
+                key=lambda b: b.energy_rate)))
+
+        else:
+            # Sorted bids in ascending order
+            return list(sorted(
+                obj.values(),
+                key=lambda b: b.energy_rate))
+
+    def calculate_match_recommend(self, bids, offers):
+        # Pay as bid first
+        # There are 2 simplistic approaches to the problem
+        # 1. Match the cheapest offer with the most expensive bid. This will favor the sellers
+        # 2. Match the cheapest offer with the cheapest bid. This will favor the buyers,
+        #    since the most affordable offers will be allocated for the most aggressive buyers.
+
+        # Sorted bids in descending order
+        sorted_bids = self.sorting(bids, True)
+
+        # Sorted offers in descending order
+        sorted_offers = self.sorting(offers, True)
+
+        already_selected_bids = set()
+        bid_offer_pairs = []
+        for offer in sorted_offers:
+            for bid in sorted_bids:
+                if bid.id not in already_selected_bids and \
+                        (offer.energy_rate - bid.energy_rate) <= \
+                        FLOATING_POINT_TOLERANCE and offer.seller != bid.buyer:
+                    already_selected_bids.add(bid.id)
+                    bid_offer_pairs.append(tuple((bid, offer, bid.energy_rate)))
+                    break
```

Comment:
Negative conditionals can sometimes make the code a bit trickier to understand. What would you think of something like the following?
```suggestion
for offer in sorted_offers:
    for bid in sorted_bids:
        if offer.seller != bid.buyer or bid.id in already_selected_bids:
            continue
       
        if (offer.energy_rate - bid.energy_rate) <= FLOATING_POINT_TOLERANCE:
            already_selected_bids.add(bid.id)
            bid_offer_pairs.append(tuple((bid, offer, bid.energy_rate)))
            break
```
What's your opinion?

Reply:
> Negative conditionals can sometimes make the code a bit trickier to understand.

Sorry, not able to think myself of cases where the negative conditional statement may make it tricky, but based on your suggestion, I will adapt it as its not much big of a change.

Reply:
It's just one common "rule", so of course we're free to decide based on the case whether it makes sense or not. I personally think it makes sense (also in daily spoken language). If you're interested, you can google for "negative conditionals programming". Here's a short explanation: https://csharpdeveloper.wordpress.com/2016/01/20/roberts-rules-of-coders-8-avoid-negative-conditionals-and-method-names/

Reply:
Thanks for the link, makes sense for future to read and understand the logic easily, accpepting your suggestion with slight modification
```suggestion
        for offer in sorted_offers:
            for bid in sorted_bids:
                if bid.id in already_selected_bids or offer.seller == bid.buyer:
                    continue
                if (bid.energy_rate - offer.energy_rate) > FLOATING_POINT_TOLERANCE:
                    already_selected_bids.add(bid.id)
                    bid_offer_pairs.append(tuple((bid, offer, bid.energy_rate)))
```